cip/CIP2016-01-26-mandatory-match.adoc

```diff
@@ -0,0 +1,72 @@
+MANDATORY MATCH (u:User {id: {userId}})
+= CIP2016-01-26 - Mandatory Match
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP proposes adding a new variant of the `MATCH` clause called `MANDATORY MATCH` that fails when it does not match at least a single row in the underlying graph.
+--
+
+toc::[]
+
+== Motivation
+
+When writing applications using Cypher, many queries will start off by looking up multiple individual nodes based on an unique id with the implied expectation of matching exactly one node. When that expectation is not fulfilled (either due to sending a wrong id or the node having been deleted) it becomes important to identify which lookup failed. Achieving  this currently is not well supported by Cypher.
```

Comment:
@cleishm: I agree, some general `REQUIRE` construct could be useful. But I think this is a different thing from what `MANDATORY MATCH` is trying to do. `MANDATORY MATCH` talks about a specific pattern having to match something, or the query will fail, whereas `REQUIRE` would talk about a required cardinality at a particular point in the query. `MANDATORY MATCH` makes the matching required for _each incoming row_, emulating that with `REQUIRE` would require the user to put the `MATCH ... REQUIRE` in a sub-query, at which point I believe `MANDATORY MATCH` is a more intuitive way of expressing the intent.